prow/gangway/gangway.go

```diff
@@ -0,0 +1,824 @@
+// purposes of fetching their job definitions. Either we duplicate a lot of the
+// same business logic, or we duplicate a lot of lower-level "infra" logic. We
+// choose to do the latter here. For an approach of duplicating the business
+// logic (which is less desirable), see the code for the "subscriber" package
+// used by the Sub component, specifically getProwJobSpec().
+type presubmitFetchers struct{}
+type postsubmitFetchers struct{}
+
+type jobFetchers interface {
+	getJobsFromStaticConfig(cfg *config.Config, orgRepo string) []interface{}
+	getJobsFromInRepoConfig(cfg *config.Config, pc *config.InRepoConfigCacheHandler, orgRepo string, baseSHAGetter config.RefGetter, headSHAGetters ...config.RefGetter) ([]interface{}, error)
+	couldRun(job interface{}, ref string) bool
+	name(job interface{}) string
+}
+
+func (jf *presubmitFetchers) getJobsFromStaticConfig(cfg *config.Config, orgRepo string) []interface{} {
+	jobs := cfg.GetPresubmitsStatic(orgRepo)
+	ret := make([]interface{}, len(jobs))
+	for i, v := range jobs {
+		ret[i] = v
+	}
+	return ret
+}
+
+func (jf *postsubmitFetchers) getJobsFromStaticConfig(cfg *config.Config, orgRepo string) []interface{} {
+	jobs := cfg.GetPostsubmitsStatic(orgRepo)
+	ret := make([]interface{}, len(jobs))
+	for i, v := range jobs {
+		ret[i] = v
+	}
+	return ret
+}
+
+func (jf *presubmitFetchers) couldRun(job interface{}, ref string) bool {
+	typedJob := job.(config.Presubmit)
+	return typedJob.CouldRun(ref)
+}
+
+func (jf *postsubmitFetchers) couldRun(job interface{}, ref string) bool {
+	typedJob := job.(config.Postsubmit)
+	return typedJob.CouldRun(ref)
+}
+
+func (jf *presubmitFetchers) name(job interface{}) string {
+	typedJob := job.(config.Presubmit)
+	return typedJob.Name
+}
+
+func (jf *postsubmitFetchers) name(job interface{}) string {
+	typedJob := job.(config.Postsubmit)
+	return typedJob.Name
+}
+
+func (jf *presubmitFetchers) getJobsFromInRepoConfig(cfg *config.Config, pc *config.InRepoConfigCacheHandler, orgRepo string, baseSHAGetter config.RefGetter, headSHAGetters ...config.RefGetter) ([]interface{}, error) {
+	jobs, err := pc.GetPresubmits(orgRepo, baseSHAGetter, headSHAGetters...)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]interface{}, len(jobs))
+	for i, v := range jobs {
+		ret[i] = v
+	}
+	return ret, nil
+}
+
+func (jf *postsubmitFetchers) getJobsFromInRepoConfig(cfg *config.Config, pc *config.InRepoConfigCacheHandler, orgRepo string, baseSHAGetter config.RefGetter, headSHAGetters ...config.RefGetter) ([]interface{}, error) {
+	jobs, err := pc.GetPostsubmits(orgRepo, baseSHAGetter)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]interface{}, len(jobs))
+	for i, v := range jobs {
+		ret[i] = v
+	}
+	return ret, nil
+}
+
+// FetchJobStruct looks at the sea of all possible Prow Job definitions and
+// selects The One that matches the details in the request.
+func (gw *Gangway) FetchJobStruct(req *CreateJobExecutionRequest) (*JobStruct, error) {
```

Comment:
Lets use the same pattern that Sub uses.  Either something similar or reusing the same code.

Reply:
For Sub, we could even consider converting to the data structures (Go structs) generated from the .proto, so that we can reuse the same business logic as in Gangway.